external/CMakeLists.txt

```diff
@@ -93,3 +93,4 @@ endif()
 add_subdirectory(db_drivers)
 add_subdirectory(easylogging++)
 add_subdirectory(randomx EXCLUDE_FROM_ALL)
+add_subdirectory(sqlite)
```

Comment:
I'm on the vendor everything train. Builds are more reliable and everyone builds the same thing that we've tested the software on.

Reply:
> I'm on the vendor everything train. Builds are more reliable and everyone builds the same thing that we've tested the software on.

This is a major problem in multiple ways.

Bundling dependencies is a sometimes necessary evil to deal with unavailable or difficult to install dependencies, cases where the upstream source has an unstable API, or dealing with cases where the software needs to be locally modified.  None of those seem to apply to a mature, stable project like sqlite3.  If our software doesn't work with particular versions of sqlite3 then we should be expressing that through a version dependency, not by forcing someone compiling the software to use a statically compiled bundled version.  There are a lot of downsides to bundling software; sometimes the pros outweigh the cons, but given the ubiquity of sqlite3 as a standard system library with a long-standing stable API (pledged by upstream to remain stable until at least 2050) I'm not seeing them at all here.

I also don't want to worry about needing to track security updates for sqlite3, boost, libunbound, openssl, and anything else that we might want to vendor, and issue new loki software releases every time there is any security update in any of them.  But *those are the implications* of vendor everywhere, and that's why it is such a software development anti-pattern in an open source project.  That additional work of needing to monitor upstream projects and issue new releases whenever something critical comes along is development time wasted with little benefit when that process is *already being done* on pretty much every sane OS (i.e. everywhere except Windows), and typically with faster reactions than we will ever be able to manage.

Furthermore "everyone builds the same thing that we've tested the software on" is simply not possible: admission into pretty much any linux distribution basically *requires* that bundled software be unbundled in the distribution package.  Downstream distributors strongly recommend that upstream projects don't forcibly bundle to save them extra work of having to unbundle when packaging for a distribution.  See [Fedora](https://fedoraproject.org/wiki/Bundled_Libraries?rd=Packaging:Bundled_Libraries), [Gentoo](https://wiki.gentoo.org/wiki/Why_not_bundle_dependencies), [Debian](https://wiki.debian.org/EmbeddedCodeCopies) policies on the topic: they all amount to the same requirement, with the same justifications.  (Since I'm maintaining the debs in such a way that they are acceptable for inclusion into Debian, that also means I'm going to have to write a patch to unbundle the debs before we release, so we won't even be getting that exact sqlite3 version requirement).

The Gentoo section on [what upstream software devs should do](https://wiki.gentoo.org/wiki/Why_not_bundle_dependencies#What_to_do_upstream) is particularly on-point: if an upstream software include a bundled dependency for convenience, please provide some way to defer to using system libraries instead.

But this *doesn't* only apply to linux distributions.  If I compile software on my system I strongly, strongly prefer that the software link to system libraries so that when I install system updates I get those updates in my compiled software without having to do anything.  You want to take that away from people compiling the source as-is for themselves, and I am not even remotely convinced by it being justified in this case.

Yet another disadvantage is that it will make the wallet api library unusable in any other project that uses sqlite because you will have symbol conflicts between the two versions.  This dependency hell doesn't happen if you dynamically link because there's only one sqlite3 library.

Bundling a static library for static build system binaries is totally fine: we already do that with all of our dependencies, like boost, but that belongs as part of the build inside `contrib/depends` that is downloaded and built on demand, not a forced bundled build in the main repository.  The change here *doesn't* merely add it only into our static builds, it also forces it on all unmodified compilations of the software, and that is the part that is not acceptable.

Quite aside from all of that, the way sqlite3 is packaged here is problematic.  First, we should not be sticking a *copy* of any subproject into our repository; rather we should either be using a submodule to fetch the upstream code (as we do with unbound, miniupnpc, unbound), or downloading it (with checksum verification) when needed as part of the build (as we do in contrib/depends, or as [lokinet does with libsodium](https://github.com/loki-project/loki-network/blob/e19714ebc1612744ef053c0a734ee655579e57a9/crypto/CMakeLists.txt#L75)).  That saves needlessly adding 9MB (!) of C code to the loki repository.

Second, the way the sqlite3 code that has been included here is the amalgamated version of the sqlite3 code.  This is problematic for licensing reasons: it makes loki core non-free software because we are not actually including the source code (in the sense of being the preferred form of modification); rather we have derived code generated from the actual source code (i.e. the preferred form of modification) that we aren't including.  That would make the loki source code non-free software by most definitions, including the Debian Free Software Guidelines and FSF.  That's unacceptable.

Reply:
My whole premise with static and vendoring is to remove friction, build ease, nothing changing behind our backs, but that is not realistic here for compliance with *nix ecosystem. Some of the arising code/workflow issues are questionable but compliance/user expectation trumps the latter.

I'll change it to dynamic linking and submodule otherwise.